lib/matplotlib/axes/_axes.py

```diff
@@ -2350,18 +2350,20 @@ def stem(self, *args, linefmt=None, markerfmt=None, basefmt=None,
             which inspired this method.
 
         """
+        if len(args) > 5:
+            raise TypeError('Too many positional arguments')
+
         # Assume there's at least one data array
         y = np.asarray(args[0])
         args = args[1:]
 
         # Try a second one
         try:
-            x, y = y, np.asarray(args[0], dtype=float)
-        except (IndexError, ValueError):
-            # The second array doesn't make sense, or it doesn't exist
+            x, *args = args
```

Comment:
That seems to be a slightly convoluted way to do `if len(args) > 1` (yes, I know, the original version is just as bad).

Moreover it is far from obvious to me that we want to drop support for passing just `y` (in which case `x` is implicitly defined as `range(len(y))`).  Yes, it's a questionable API, but it's also present in `plot()` and I don't think it makes sense to remove it in a piecemeal fashion; instead this should be done in a consistent manner.

Of course, there's the additional complication of categorical data (although I think categorical y's in a stem plot don't make much sense either).

TLDR: I don't think the API change is as benign as the PR suggests, and warrants more discussion.

Reply:
Agreed, a check `if len(args) > 1` is preferable to this `try` logic.

Support for `y`-only is not dropped. See the except clause.